### ISSUE TITLE
Problem with db.login()

### ISSUE THREAD

**Issue by choemi** [nolanlawson/pouchdb-authentication#51]
I created the following test code:

```javascript
var user = {
  name: 'user',
  password: 'test'
};

var pouchOpts = {
  skipSetup: true
};

var ajaxOpts = {
  ajax: {
    headers: {
      Authorization: 'Basic ' + window.btoa(user.name + ':' + user.password)
    }
  }
};

var db = new PouchDB('https://pouchdb-auth.iriscouch.com/auth-demo', pouchOpts);
var err;


function userLogIn() {
	db.login(user.name, user.password, ajaxOpts).then(function() {
	  return db.allDocs();
	}).then(function(docs) {
	  console.log(docs);
	}).catch(function(error) {
	  console.error(error);
	});
}
```
My test HTML file has something like this in it

```html
<button class="btn btn-primary" onclick="userLogIn()">userLogIn()</button><br>
```

CORS is enabled on IrisCouch (credentials, headers, methods and origins are defined) and require_valid_user = true, because I need the user to login in first before doing anything with my db.

But when I am calling the userLogIn() by pressing the button, I still get the old login windows from the browser. When I enter my credentials, the login works fine. But I'd like to build my own login screen. Shouldn't this be working with ajaxOpts? Or do I have to add anything else / configure something more on IrisCouch?

Thanks for helping me in advance. This is my first project with PouchDB, so I might do something wrong...

**Comment by nolanlawson** [comment]
No prob; it's very confusing to get authentication working, because there are so many ways to do it... :)

In this case, it looks like your error is this line:

    Authorization: 'Basic ' + window.btoa(user.name + ':' + user.password)

You don't need to add this. `db.login()` uses cookie authentication, whereas that line is using "basic" authentication (aka "regular HTTP" authentication), which is what shows the popup window.

Let me know if that helps!

**Comment by choemi** [comment]
Thanks for your fast reply.
So this would mean it could just remove the ajaxOpts like?
```javascript
 db.login(user.name, user.password).then(function()
```

But this doesn't solve the problem, the window still pops up. When keeping ajaxOpts in db.login() and just removing Authorization, same problem appears. Anyway, I guess it's same and it doesn't matter if removing it from db.login or just from ajaxOpts.

```js
var ajaxOpts = {
  ajax: {
    headers: {
    }
  }
};
```

Any other ideas?

**Comment by nolanlawson** [comment]
That's really, really weird. If you check out the codebase for pouchdb-authentication and run the tests yourself against CouchDB, you'll see that they login/logout without any popup window. You can verify with:

    git clone https://github.com/nolanlawson/pouchdb-authentication.git
    cd pouchdb-authentication
    npm install && npm run dev

Then open the tests up in a browser and it will try to login/logout to localhost:5984.

**Comment by choemi** [comment]
I tried the following:
```js
git clone https://github.com/nolanlawson/pouchdb-authentication.git
cd pouchdb-authentication
npm install
```
First I got an error Can't find Python executable "python", you can set the PYTHON en variable.
So I read something in the internet to fix this (http://stackoverflow.com/questions/21365714/nodejs-error-installing-with-npm)
I just executed
```js
npm install -g node-gyp
```
and I don't see the error again. Do I really have to install MS Visual Studio C++ and Python on Windows to make it work?

Now, when running
```js
npm run dev
```
I get another error
Can't open lib/*.js

If it helps, I could send you the logs (npm-debug.log)

Anyway, I will try it later on a MacBook, maybe it works there...

**Comment by nolanlawson** [comment]
Yeah, our build process is really flakey on Windows. What I'm really trying to demonstrate is just that it works&trade;. :)

Here is a demo HTML site. Download it (or `git clone` it) and run it with `python -m SimpleHTTPServer` or some other HTTP server and tell me if it works (assuming your CouchDB is as `localhost:5984`): https://gist.github.com/nolanlawson/ac387be0722dad3a207c

**Comment by choemi** [comment]
So, I finally tried it on a MacBook, but unfortunately no success.

With your HTML demo site, I got the following error after running SimpleHTTPServer and going to http://localhost:8000/ CouchDB runs on 5984, CORS enabled.
```
http
authentication
"before each" hook ‣
{"status":500,"name":"unknown_error","message":"Database encountered an unknown error"}db = new PouchDB(dbName);
return db;
```

I also tried to clone the whole bundle:
````
git clone https://github.com/nolanlawson/pouchdb-authentication.git
cd pouchdb-authentication
npm install && npm run dev
```
and went to `http://127.0.0.1:8002/test/`

Then I got 
```
http
authentication
Test signup ‣
Test signup conflict51ms ‣
Test bad signup args ‣
Test bad signup args 2 ‣
Test login/logout ‣
TypeError: Cannot read property 'should' of null
    at http://127.0.0.1:8002/test/test-bundle.js:25054:25return db.signup('aquaman', 'sleeps_with_fishes').then(function (res) {
  return db.getSession();
}).then(function (res) {
  should.equal(res.userCtx.name, null);
  return db.login('aquaman', 'sleeps_with_fishes');
}).then(function (res) {
  res.ok.should.equal(true);
  return db.getSession();
}).then(function (res) {
  res.userCtx.name.should.equal('aquaman');
  return db.logout();
}).then(function () {
  return db.getSession();
}).then(function (res) {
  should.equal(res.userCtx.name, null);
});
Test metadata ‣
Test wrong user for getUser ‣
AssertionError: expected 'AssertionError' to equal 'not_found'
Test change password42ms
```

Any other ideas?

**Comment by nolanlawson** [comment]
Which browser are you using? I just tested in Chrome on a Mac, and those steps work fine for me. I also have a CouchDB with CORS running at localhost:5984. `http://127.0.0.1:8002/test/` passes with 100% success.

**Comment by nolanlawson** [comment]
BTW that "500 unknown error" means there was some OTHER error that caused it. You might have to dig through the console or the Network tab to see it; sorry about that...

**Comment by SinanGabel** [comment]
I use it on MacBook but so far I have only used it through browser (i.e. not nodejs) by adding the <script> in the index.html file, and using a local instance of CouchDB.

`<script type="text/javascript" src="PATH_TO_SOURCE/pouchdb-4.0.0.min.js"></script>`

It works with all the major browsers.

I.e. an idea could be to get it up and running through browser first, and then when that works try nodejs again.

**Comment by choemi** [comment]
I am using also Chrome. For the demo in shows me in the console:
`test-bundle.js:19754 GET http://localhost:2022/testdb/?_nonce=1441031614430 net::ERR_CONNECTION_REFUSED`
Before, running the server and the test file, I have created a db called `testdb` because in the test.js file, there is written:
`process.env.TEST_DB = 'http://localhost:2022/testdb';`
Anything else to add?
Btw: why is it using port 2022 and not 5984?


For the bundle, I get the following error:
`Failed to load resource: the server responded with a status of 404 (Not Found)`
`http://localhost:2022/testdb/_local/_pouch_dependentDbs?_nonce=1441031839657`
and then in browser window itself when opening the URL above
`{"error":"not_found","reason":"no_db_file"}`

Do I have to make anything on the CouchDB except enabling CORS and let it run. I am still in Admin party.

I am really wondering if it is a problem with my CouchDB? Can I change somewhere in the test file the server to IrisCouch? Would this be in the test.js?  `process.env.TEST_DB = 'http://localhost:2022/testdb'` meaning replacing localhost just with the IrisCouch db url?

**Comment by nolanlawson** [comment]
2022 is because the test uses a proxy. As for the 404, that is totally normal.

It definitely sounds like you have a problem with your CouchDB. To run the tests against your own IrisCouch, yeah, you would need to edit the code to make it point somewhere other than localhost:2022.

**Comment by tohagan** [comment]
I'm getting very similar issues (same 500 error) that I've failed to resolve. I have no proxy - just CouchDb 1.6.1 running on Win7.   Tried Cloundant (but it stalls with 400 error) .

* Very carefully checked that I have original .ini files with minimal changes. 

* Tried many configurations over 2 weeks with CORS, credentials true/false, All imaginable CORS, origins *  or specific URL (matching that logged by CouchDb),  All headers, skipSetup: true/false.

Basically I can't get db.login() to work.

**Comment by nolanlawson** [comment]
I'm going to close this unless somebody can provide steps to reproduce. For anybody still experiencing this, please check out the code, run the tests, and tell me if they don't work. For me, they work.

Steps to test:

```
git clone https://github.com/nolanlawson/pouchdb-authentication.git
cd pouchdb-authentication
npm install
npm run dev
```

**Comment by DQvsRA** [comment]
That is what i have after running npm run dev

![screenshot](https://cloud.githubusercontent.com/assets/442111/22995484/5cf29cda-f3cb-11e6-887a-9a00a738fe4a.png)

**Comment by nolanlawson** [comment]
@DQvsRA Did you run `add-cors-to-couchdb`?

**Comment by lukasborawski** [comment]
@nolanlawson i have the same issue on Safari, on Chrome it works correctly :(
